sources/files.go

```diff
@@ -0,0 +1,192 @@
+			if os.IsPermission(err) {
+				// This seems to only fail on directories at this stage.
+				logger.Warn().Msgf("skipping directory: permission denied")
+				return filepath.SkipDir
+			}
+			logger.Error().Msgf("skipping directory: %s", err)
+			return nil
+		}
+
+		info, err := d.Info()
+		if err != nil {
+			if d.IsDir() {
+				logger.Error().Msgf("skipping directory: could not get info: %s", err)
+				return filepath.SkipDir
+			}
+			logger.Error().Msgf("skipping file: could not get info: %s", err)
+			return nil
+		}
+
+		if !d.IsDir() {
+			// Empty; nothing to do here.
+			if info.Size() == 0 {
+				logger.Debug().Msg("skipping file: size=0")
+				return nil
+			}
+
+			// Too large; nothing to do here.
+			if s.MaxFileSize > 0 && info.Size() > int64(s.MaxFileSize) {
+				logger.Warn().Msgf(
+					"skipping file: too large: max_size=%dMiB, size=%dMiB",
+					s.MaxFileSize/1000000, info.Size()/1000000,
+				)
+				return nil
+			}
+		}
+
+		// set the inital scan target values
+		if d.Type()&fs.ModeSymlink != 0 {
+			if !s.FollowSymlinks {
+				logger.Debug().Msg("skipping symlink: follow symlinks disabled")
+				return nil
+			}
+			realPath, err := filepath.EvalSymlinks(path)
+			if err != nil {
+				logger.Error().Msgf("skipping symlink: could not evaluate: %s", err)
+				return nil
+			}
+			realPathFileInfo, _ := os.Stat(realPath)
+			if realPathFileInfo.IsDir() {
+				logger.Debug().Msgf("skipping symlink: target is directory: target=%q", realPath)
+				return nil
+			}
+			scanTarget = ScanTarget{
+				Path:    realPath,
+				Symlink: path,
+			}
+		}
+
+		// handle dir cases (mainly just see if it should be skipped
+		if info.IsDir() {
+			if s.shouldSkip(path) {
+				logger.Debug().Msg("skipping directory: global allowlist item:")
+				return filepath.SkipDir
+			}
+			if info.Name() == ".git" {
+				// TODO: Add this to the config allowlist, instead of hard-coding it.
+				logger.Debug().Msg("skipping directory: .git directory always skipped")
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		if s.shouldSkip(path) {
+			logger.Debug().Msg("skipping file: global allowlist item:")
+			return nil
+		}
+
+		if err := yield(scanTarget, nil); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (s *Files) Fragments(yield FragmentsFunc) error {
+	var wg sync.WaitGroup
+
+	return s.scanTargets(func(scanTarget ScanTarget, err error) error {
+		wg.Add(1)
+
+		s.Sema.Go(func() error {
+			defer wg.Done()
+			logger := logging.With().Str("path", scanTarget.Path).Logger()
+			logger.Trace().Msg("scanning path")
+
+			f, err := os.Open(scanTarget.Path)
+			if err != nil {
+				if os.IsPermission(err) {
+					logger.Warn().Msg("skipping file: permission denied")
+				}
+				return nil
+			}
+			defer f.Close()
+
+			// Convert this to a file source
+			file := File{
+				Content: f,
+				Path:    scanTarget.Path,
+				Symlink: scanTarget.Symlink,
+			}
+
+			return file.Fragments(yield)
+		})
+
+		wg.Wait()
```

Comment:
Won't this run each target serially, compared to `DetectFiles`?

Reply:
Yerp! Put it in the wrong spot.